tempesta_fw/sock_clnt.c

```diff
@@ -45,28 +45,62 @@
  * ------------------------------------------------------------------------
  */
 
+/**
+ * TfwConnection extension for client sockets.
+ *
+ * @conn		- The base structure. Must be the first member.
+ * @keepalive_timer	- The keep-alive timer for the connection.
+ * @timeout		- The keep-alive timeout for the connection.
+ *
+ */
+typedef struct {
+	TfwConnection		conn;
+	struct timer_list	keepalive_timer;
+} TfwCliConnection;
+
 static struct kmem_cache *tfw_cli_conn_cache;
+static int tfw_cli_cfg_ka_timeout = -1;
+
+static void
+tfw_sock_cli_keepalive_timer_cb(unsigned long data)
+{
+	TfwCliConnection *cli_conn = (TfwCliConnection *)data;
+	TFW_DBG("Client timeout end\n");
+	if (ss_close(cli_conn->conn.sk) == SS_OK) {
+		tfw_connection_unlink_from_sk(cli_conn->conn.sk);
```

Comment:
Copy&Paste could be dangerous. I think this code should be moved to a separate function called from here and from`tfw_sock_clnt_drop()`. Also, comments to this code should be moved to the new function from `tfw_sock_clnt_drop()`.